template-generator/create_template.py

```diff
@@ -0,0 +1,473 @@
+import argparse
+import re
+import yaml
+import openpyxl
+import xlsxwriter
+from bs4 import BeautifulSoup, Tag
+from copy import deepcopy
+import inflect
+
+
+def parse_arguments():
+    parser = argparse.ArgumentParser(
+        description="""
+        USDM Comformance Rules Test Data Template Generator.
+        Reads the Enterprise Architect XMI export file containing definition
+        of the USDM UML model (specified in the -x option), the USDM
+        Controlled Terminology Excel file (specified in the -c option), and the
+        USDM API specification in YAML format (specified in the -a option) and
+        creates an Excel template file (optionally specified in the -o option)
+        for test data entry."""
+    )
+    parser.add_argument(
+        "-x",
+        "--xmi_file",
+        help="USDM XMI export file "
+        + "(e.g.,<DDFversion>/Deliverables/UML/USDM_UML.xmi)",
+        required=True,
+    )
+    parser.add_argument(
+        "-c",
+        "--ct_file",
+        help="USDM CT Excel file "
+        + "(e.g., <DDF version>/Deliverables/CT/USDM_CT.xlsx)",
+        required=True,
+    )
+    parser.add_argument(
+        "-a",
+        "--api_spec",
+        help="USDM API specification YAML file. "
+        + "(e.g., <DDF version>/Deliverables/API/USDM_API.yaml)",
+        required=True,
+    )
+    parser.add_argument(
+        "-o",
+        "--output_file",
+        help="[Optional] Specifies output file Excel file. "
+        + "Default is ./USDM_Test_Data_Template.xlsx",
+        default="USDM_Test_Data_Template.xlsx",
+    )
+    args = parser.parse_args()
+    return args
+
+
+def replace_deep(data, a, b):
+    if isinstance(data, str):
+        return data.replace(a, b)
+    elif isinstance(data, dict):
+        return {k: replace_deep(v, a, b) for k, v in data.items()}
+    elif isinstance(data, list):
+        return [replace_deep(v, a, b) for v in data]
+    else:
+        return data
+
+
+def get_properties(
+    entName: str, cls: Tag, prps: list, prefix: str = None, lclsName: str = None
+):
+    if cls.generalization:
+        gclsId = cls.generalization["general"]
+        gcls = usdmxmi.find("packagedElement", attrs={"xmi:id": gclsId})
+        gclsName = gcls["name"]
+        get_properties(entName, gcls, prps, prefix, gclsName)
+    for prp in (
+        x
+        for x in cls.find_all("ownedAttribute", attrs={"xmi:type": "uml:Property"})
+        if x.has_attr("name")
+    ):
+        prps[0] += [".".join((prefix, prp["name"])) if prefix else prp["name"]]
```

Comment:
If we use space instead of `.` notation for the column delimiters, we can add cell wrapping and it should wrap nicely over multiple rows.

Reply:
How would this align with the naming of columns in the pandas(?) dataframe that would be created from JSON input files by the engine?  Are spaces allowed there too?

Reply:
After discussion with Rebecca, we decided that we both prefer the dot notation, but agree that wrapping the column headers would be better.

Reply:
Wrapping added for header row.  Also increased the column width a bit.